image_classification/FF_Only/ffonly.py

```diff
@@ -0,0 +1,250 @@
+        self.num_features = self.embed_dim = embed_dim
+
+        # Patch embedding
+        self.patch_embed = PatchEmbed(patch_size=patch_size, in_chans=in_chans, embed_dim=embed_dim)
+
+        # Class token
+        self.cls_token = self.create_parameter(shape=[1, 1, embed_dim], default_initializer=trunc_normal_)
+        self.add_parameter("cls_token", self.cls_token)
+
+        # Positional encoding
+        if positional_encoding == 'learned':
+            height, width = self.learned_positional_encoding_size = learned_positional_encoding_size
+            self.pos_encoding = LearnedPositionalEncoding(height, width, embed_dim)
+        else:
+            raise NotImplementedError('Unsupposed positional encoding')
+        self.pos_drop = nn.Dropout(p=drop_rate)
+
+        # Stochastic depth
+        dpr = [x.item() for x in paddle.linspace(0, drop_path_rate, depth)]
+        self.blocks = nn.LayerList([
+            block_cls(dim=embed_dim, num_heads=num_heads, mlp_ratio=mlp_ratio, qkv_bias=qkv_bias, qk_scale=qk_scale,
+                      drop=drop_rate, attn_drop=attn_drop_rate, drop_path=dpr[i], norm_layer=norm_layer, num_tokens=1 + (224 // patch_size)**2)
+            for i in range(depth)])
+        self.norm = norm_layer(embed_dim)
+
+        # Classifier head
+        self.head = nn.Linear(embed_dim, num_classes) if num_classes > 0 else Identity()
+        self.apply(self._init_weights)
+
+    def _init_weights(self, m):
+        if isinstance(m, nn.Linear):
+            trunc_normal_(m.weight)
+            if isinstance(m, nn.Linear) and m.bias is not None:
+                zeros_(m.bias)
+        elif isinstance(m, nn.LayerNorm):
+            zeros_(m.bias)
+            ones_(m.weight)
+
+
+    def forward_features(self, x):
+
+        # Patch embedding
+        B, C, H, W = x.shape  # B x C x H x W
+        x = self.patch_embed(x)  # B x E x H//p x W//p
+
+        # Positional encoding
+        # NOTE: cls_pos_embed for compatibility with pretrained models
+        cls_pos_embed, pos_embed = self.pos_encoding(x)
+
+        # Flatten image, append class token, add positional encoding
+        cls_tokens = self.cls_token.expand([B, -1, -1])
+        x = x.flatten(2)  # flatten
+        x = swapdim(x , 1, 2)
+        x = paddle.concat((cls_tokens, x), axis=1)  # class token
+        pos_embed = pos_embed.flatten(2)            # flatten
+        pos_embed = swapdim(pos_embed, 1, 2)
+        pos_embed = paddle.concat([cls_pos_embed, pos_embed], axis=1)  # class pos emb
+        x = x + pos_embed
+        x = self.pos_drop(x)
+
+        # Transformer
+        for blk in self.blocks:
+            x = blk(x)
+
+        # Final layernorm
+        x = self.norm(x)
+        return x[:, 0]
+
+    def forward(self, x):
+        x = self.forward_features(x)
+        x = self.head(x)
+        return x
+        
+
+def build_ffonly(config):
+    model = LinearVisionTransformer(
+        patch_size = config.MODEL.MIXER.PATCH_SIZE, 

```

Comment:
Please use pylint to check and fix errors, warnings or conventions,
Please add docstring for classes and methods.

Reply:
This code raise errors when using Multi-GPU for evaluation:
```
	Traceback (most recent call last):
  File "main_multi_gpu.py", line 362, in <module>
    main()
  File "main_multi_gpu.py", line 358, in main
    dist.spawn(main_worker, args=(dataset_train, dataset_val, ), nprocs=config.NGPUS)
  File "/opt/conda/envs/py36/lib/python3.6/site-packages/paddle/distributed/spawn.py", line 496, in spawn
    while not context.join():
  File "/opt/conda/envs/py36/lib/python3.6/site-packages/paddle/distributed/spawn.py", line 296, in join
    self._throw_exception(error_index)
  File "/opt/conda/envs/py36/lib/python3.6/site-packages/paddle/distributed/spawn.py", line 314, in _throw_exception
    raise Exception(msg)
Exception:

----------------------------------------------
Process 0 terminated with the following error:
----------------------------------------------

Traceback (most recent call last):
  File "/opt/conda/envs/py36/lib/python3.6/site-packages/paddle/distributed/spawn.py", line 245, in _func_wrapper
    result = func(*args)
  File "/workspace/ppvit_github/PaddleViT/image_classification/FF_Only/main_multi_gpu.py", line 303, in main_worker
    debug_steps=config.REPORT_FREQ)
  File "/workspace/ppvit_github/PaddleViT/image_classification/FF_Only/main_multi_gpu.py", line 172, in validate
    dist.all_reduce(loss)
  File "/opt/conda/envs/py36/lib/python3.6/site-packages/paddle/distributed/collective.py", line 446, in all_reduce
    tensor, 'use_calc_stream', use_calc_stream, 'ring_id', ring_id)
ValueError: (InvalidArgument) Communicator of ring id 0 has not been initialized.
  [Hint: Expected comm_map_.count(ring_id) > 0, but received comm_map_.count(ring_id):0 <= 0:0.] (at /paddle/paddle/fluid/platform/collective_helper.h:103)
  [operator < c_allreduce_sum > error]
```